src/utilities/leetcode.ts

```diff
@@ -0,0 +1,342 @@
+ * @param titleSlug The title slug of the problem
+ * @returns The last submission ID
+ */
+export const fetchLastAcceptedId = async (
+  session: string,
+  titleSlug: string
+): Promise<number> => {
+  // Define the query
+  const query = `
+    query submissionList($offset: Int!, $limit: Int!, $lastKey: String, $questionSlug: String!, $lang: Int, $status: Int) {
+      questionSubmissionList(
+        offset: $offset
+        limit: $limit
+        lastKey: $lastKey
+        questionSlug: $questionSlug
+        lang: $lang
+        status: $status
+      ) {
+        lastKey
+        hasNext
+        submissions {
+          id
+        }
+      }
+    }
+  `;
+
+  // Send the request
+  const response = await fetch(ENDPOINT, {
+    headers: {
+      "content-type": "application/json",
+      cookie: `LEETCODE_SESSION=${session}`,
+    },
+    body: JSON.stringify({
+      query,
+      variables: {
+        offset: 0,
+        limit: 20,
+        questionSlug: titleSlug,
+        status: 10,
+      },
+      operationName: "submissionList",
+    }),
+    method: "POST",
+  });
+
+  // Validate the response
+  if (!response.ok) {
+    throw new Error("fetchLastAcceptedId, invalid status");
+  }
+  const { data } = await response.json();
+  if (data === null || data.questionSubmissionList === null) {
+    throw new Error("fetchLastAcceptedId, empty response");
+  }
+
+  // Parse the response
+  const { submissions } = data.questionSubmissionList;
+  return parseInt(submissions[0].id);
+};
+
+/**
+ * Fetch the history of submissions from LeetCode.
+ * @param session The session cookie
+ * @param reprotFn The function to report progress (i.e. completion of m out of n)
+ * @returns The progress and submission history
+ */
+export const fetchHistory = async (
+  session: string,
+  reprotFn: (m: number, n: number) => void
+): Promise<[Progress[], Submission[]]> => {
+  // Fetch the progress at each page
+  const progress: Progress[] = [];
+  var [pageNo, hasMore] = [1, true];
+  while (hasMore) {
+    const response = await retry(() => fetchProgressAtPage(session, pageNo));
+    var [hasMore, items] = response;
+    progress.push(...items);
+    pageNo++;
+  }
+
+  // Fetch the submission details for each item
+  const history: Submission[] = [];
+  for (let i = 0; i < progress.length; i += 4) {
+    const promises = progress.slice(i, i + 4).map(async (item, idx) => {
+      const fn1 = () => fetchLastAcceptedId(session, item.titleSlug);
+      const acceptedId = await retry(fn1);
+      await new Promise((resolve) => setTimeout(resolve, Math.random() * 1000));
+      const fn2 = () => fetchSubmissionById(session, acceptedId);
+      const submission = await retry(fn2);
+      await new Promise((resolve) => setTimeout(resolve, Math.random() * 1000));
+      return submission;
+    });
+    const resolves = await Promise.all(promises);
+    history.push(...resolves);
+    reprotFn(history.length, progress.length);
```

Comment:
Let's round these to the nearest integer.

Reply:
Fixed. I truncate the percentage to at most two decimal places.